// Copyright 2017 The Forseti Security Authors. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package playground;

service Playground {
  rpc Ping (PingRequest) returns (PingReply) {}
  rpc SetIamPolicy (SetIamPolicyRequest) returns (SetIamPolicyReply) {}

  rpc AddGroupMember (AddGroupMemberRequest) returns (AddGroupMemberReply) {}
  rpc DeleteGroupMember (DeleteGroupMemberRequest) returns (DeleteGroupMemberReply) {}

  rpc AddRole (AddRoleRequest) returns (AddRoleReply) {}
  rpc DeleteRole (DeleteRoleRequest) returns (DeleteRoleReply) {}
}

message AddRoleRequest {
  string role_name = 1;
  repeated string permissions = 2;
}

message AddRoleReply {
}

message DeleteRoleRequest {
  string role_name = 1;
}

message DeleteRoleReply {
}

message AddGroupMemberRequest {
  string member_type_name = 1;
  repeated string parent_type_names = 2;
}

message AddGroupMemberReply {
}

message DeleteGroupMemberRequest {
  string member_name = 1;
  string parent_name = 2;
  bool only_delete_relationship = 3;
}

message DeleteGroupMemberReply {
}

message PingRequest {
  string data = 1;
}

message PingReply {
  string data = 1;
}

message SetIamPolicyRequest {
  string resource = 1;
  Policy policy = 2;
}

message SetIamPolicyReply {
  bool accepted = 1;
}

message Policy {
  repeated Binding bindings = 1;
  string etag = 2;
}

message Binding {
  string role = 1;
  repeated string members = 2;
}
